File: sources/scalac/transformer/matching/LeftTracerInScala.java
package scalac.transformer.matching ;

import scalac.*;
import scalac.ast.*;
import scalac.symtab.*;
import Tree.*;
import scalac.util.Name;

import scalac.transformer.TransMatch.Matcher ;

import java.util.* ;

import ch.epfl.lamp.util.Position;

public class LeftTracerInScala extends Autom2Scala {

      Tree selector;

      /** symbol of the accumulator ( scala.SequenceList )
       */
      Symbol accumSym;


      Type _accumType( Type elemType ) {
            return new Type.TypeRef( defs.SCALA_TYPE,
                                     cf.seqTraceSym(),
                                     new Type[] { elemType });
      }


      Matcher _m ;
      public LeftTracerInScala( DetWordAutom dfa,
                                Type elementType,
                                Matcher m,
                                CodeFactory cf ) {
            // ignore clazzOwner!
            super( dfa, elementType, m.owner, cf );
            this._m = m;
            this.selector = m.selector;
            helpMap = new HashMap();
            helpVarDefs = new Vector();

      }


      // SAME AS IN RightTracerInScala,
      // todo create common abstract superclass Tracer/TransducerInScala,
      // move WordAutom down.
      HashMap helpMap ;
      Vector  helpVarDefs;

      Symbol makeHelpVar( Symbol realVar ) {
	  Symbol helpVar = new TermSymbol( //Kinds.VAR,
                                             pos,
                                             cf.fresh.newName( realVar.name
                                                               .toString() ),
                                             owner,
                                             0)
                  .setType( cf.SeqListType( elementType ) ) ;

            helpMap.put( realVar, helpVar );

            Tree varDef = gen.ValDef(helpVar, cf.newSeqNil( elementType ));
            //((ValDef) varDef).kind = Kinds.VAR;
            helpVarDefs.add( varDef );
            return helpVar;
      }

      Symbol makeHelpVarSEQ( Tree pat ) {
            String  helpName = String.valueOf( pat.hashCode() ); //wicked, in'it ?
            Symbol helpVar =
		new TermSymbol( /*Kinds.VAR, */pos,
                                  cf.fresh.newName(Name.fromString( helpName )),
                                  owner,
                                  0)
                  .setType( pat.type() ) ;

            ValDef varDef = (ValDef) gen.ValDef( helpVar,
                                                 cf.ignoreValue( pat.type() ));
            //varDef.kind = Kinds.VAR;
            helpVarDefs.add( varDef );
            return helpVar;
      }

      // SAME AS RIGHT
      Tree refHelpVar( Symbol realVar ) {
            Symbol hv = (Symbol)helpMap.get( realVar );
            assert hv != null : realVar;
            return gen.Ident( Position.NOPOS, hv );
      }

      // SAME AS RIGHT
      Tree assignToHelpVar( Symbol realVar, Tree rhs ) {
            Tree hv = refHelpVar( realVar );
            return gen.Assign( hv, rhs );
      }

      void handleVars( Vector freeVars ) {
            for(Iterator it = freeVars.iterator(); it.hasNext(); ) {
                  makeHelpVar( (Symbol) it.next() );
            }
      }

      Tree bindVar(Symbol realVar) {
            Tree hv = refHelpVar( realVar );
            System.out.println("realVar.name "+realVar.name+" type:"+realVar.type());
            if( realVar.type().isSameAs( elementType ))
                  return gen.ValDef( realVar, cf.SeqList_head( hv ));
            else
                  return gen.ValDef( realVar, hv);

      }


      /** returns a Tree whose type is boolean.
       *  now we care about free vars
       */
      /*
      Tree handleBody( HashMap helpMap ) {
            Tree res[] = new Tree[ helpMap.keySet().size() + 1 ];
            int j = 0;
            for( Iterator it = helpMap.keySet().iterator(); it.hasNext(); ) {
                  Symbol vsym = (Symbol) it.next();
                  Symbol hv   = (Symbol) helpMap.get( vsym );
                  hv.type( cf.SeqListType( elementType ) ) ;
                  Tree refv   = gen.Ident(Position.NOPOS, vsym);
                  Tree refhv  = gen.Ident(Position.NOPOS, hv);
                  res[ j++ ] = gen.Assign( refhv, refv );
            }

            res[ j ] = super.handleBody( freeVars ); // just `true'

            return cf.Block(Position.NOPOS, res, res[j].type() );
      }
*/
      protected void initializeSyms() {
            funSymName = "leftTracer";

            nestedMap = new HashMap();

            super.initializeSyms();

            this.accumSym = newParam("accum")                    // accumulator
                  .setType( _accumType( elementType ));

            this.funSym
                  .setType( new Type.MethodType( new Symbol[] {
                        accumSym, iterSym, stateSym},
                                              _accumType( elementType )));

            // 2 do: rename switchresultsym to something else...

            this.resultSym = new TermSymbol( //Kinds.VAR,
                                                   pos,
                                                   cf.fresh.newName("trace"),
                                                   owner,
                                                   0 )
                  .setType( _accumType( elementType ) ) ;

      }

      /** do the translation
      public void translate() {
            initializeSyms();
            Tree tb = code_body();
            theDefDef = gen.DefDef( this.funSym,
                                    tb );
      }
       */

      // should throw an exception here really, e.g. MatchError
      Tree code_fail() {

            return gen.Ident( accumSym.pos, accumSym );

      }



      /** returns translation of transition with label from i.
       *  returns null if there is no such transition(no translation needed)
       */
      Tree code_delta( int i, Label label ) {
            Integer target = dfa.delta( i, label );
            /*
            System.out.println("LeftTracer:calling dfa.delta("+i+","+label+")");
            System.out.println("result: "+target);
            */
            if( target == null )
                  return null;

            // (optimization) that one is a dead state (does not make sense for tracer)
            /*
              if( target == dfa.nstates - 1 )
            return code_fail();
            */
            Tree newAcc = cf.newSeqTraceCons(new Integer(i),
                                             currentElem(),
                                             _ref( accumSym ));

            return callFun( new Tree[] { newAcc , _iter(), cf.Int( target )} );
      }


      /** return code for state i of the dfa SAME AS IN SUPER, ONLY SINK IS GONE
       */
      Tree code_state( int i, Tree elseBody ) {

            Tree runFinished; // holds result of the run
            int  finalSwRes;

            runFinished = run_finished( i );

            Tree stateBody ; // action(delta) for one particular label/test

            // default action (fail if there is none)

            stateBody = code_delta( i, Label.DefaultLabel);

            if( stateBody == null )
                  stateBody = code_fail();
            // transitions of state i

            HashMap trans = ((HashMap[])dfa.deltaq)[ i ];

            for( Iterator labs = ((HashMap)dfa.deltaq( i )).keySet().iterator();
                 labs.hasNext() ; ) {
                  Object label = labs.next();
                  Integer next = (Integer) trans.get( label );


                  Tree action = code_delta( i, (Label) label );

                  if( action != null ) {

                        stateBody = cf.If( _cur_eq( _iter(), (Label) label ),
                                         action,
                                         stateBody);
                  }
            }

            return wrapStateBody( stateBody,
                                  elseBody,
                                  runFinished,
                                  i );

      }

      Tree[] getTrace() {

            initializeSyms();
            Tree tb = code_body();
            theDefDef = gen.DefDef( this.funSym,
                                    tb );

            Vector v = new Vector();

            v.addAll( helpVarDefs );

            //
            // `def leftTracer(...) = ...'                 the function definition
            v.add( theDefDef );

            Tree emptyAcc    = cf._seqTraceNil( elementType );

            // the valdef is needed, because passing emptyAcc as a parameter
            //   results in a ClassCastException at runtime (?!)

            Symbol emptyAccSym = new TermSymbol( //Kinds.VAR,
                                                 pos,
                                                 cf.fresh.newName("acc"),
                                                 owner,
                                                 0 )
                  .setType( _accumType( elementType ) ) ;

            // `val acc = SeqNil[ elementType ]'                 init accumulator
            v.add( gen.ValDef( pos, emptyAccSym, emptyAcc) );

            Tree run = callFun( new Tree[] {
                  gen.Ident( pos, emptyAccSym ),
                  cf.newIterator( selector ),
                  cf.Int( 0 )  });

            run = gen.ValDef( Position.NOPOS, resultSym, run );

            v.add( run );

            // vars...
            for( Iterator it = helpMap.keySet().iterator(); it.hasNext(); ) {
                  v.add( bindVar( (Symbol) it.next()) );
            }

            /* IF YOU NEED DEBUG OUTPUT AT RUNTIME
            v.add( cf.debugPrintRuntime( "the trace is" ) );
            v.add( cf.debugPrintRuntime( gen.Ident( pos, resultSym ) ) );
            v.add( cf.debugPrintNewlineRuntime( "" ) );
            */

            Tree res[] = new Tree[ v.size() ];
            int j = 0;
            for( Iterator it = v.iterator(); it.hasNext(); )
                  res[ j++ ] = (Tree) it.next();

            _m.varMap = nestedMap;

            return res;

      }

      public HashMap nestedMap;

      // calling the AlgebraicMatcher here
      Tree _cur_match( Tree pat ) {
            //System.out.println("calling algebraic matcher on type:"+pat.type);

            Matcher m = new Matcher( funSym,
                                     currentElem(),
                                     defs.BOOLEAN_TYPE );

            if( CollectVariableTraverser.containsBinding( pat )) {
                  switch( pat ) {
                  case Sequence(Tree[] pats):
                        //System.out.println("ouch! v Left");
                        Symbol hv = makeHelpVarSEQ( pat );
                        nestedMap.put( pat, hv );
                        Tree stm  = gen.Assign( gen.Ident(0, hv), currentElem() );
                        m.stms = new Tree[2];
                        m.stms[0] = stm;
                        m.stms[1] = gen.mkBooleanLit(Position.NOPOS, true);
                        return cf.Block( 0, m.stms, m.stms[1].type() );
                  }
            }

            FreshVariableTraverser fv = new FreshVariableTraverser(pos,
                                                                   owner,
                                                                   cf.fresh);
            fv.traverse( pat );
            HashMap helpMap = fv.helpMap;
            //System.out.println("varMap: "+helpMap );

            m.varMap = fv.helpMap;

            //replaceVars( pat );

            am.construct( m, new CaseDef[] {
                  (CaseDef) cf.make.CaseDef( pat.pos,
                                             pat,
                                             Tree.Empty,
                                             handleBody( fv.helpMap )),
                        (CaseDef) cf.make.CaseDef( pat.pos,
                                                   cf.make.Ident(pat.pos, WILDCARD_N)
                                                   .setSymbol( Symbol.NONE )
                                                   .setType(pat.type()),
                                                   Tree.Empty,
                                                   gen.mkBooleanLit(Position.NOPOS, false)) }/*,
											       false */
                          );
            Tree res = am.toTree().setType( defs.BOOLEAN_TYPE );
            //System.out.println("freeVars: "+freeVars);

            return res;
      }


      /** return the accumulator + last state
       */
      Tree run_finished( int state ) {
            return cf.newSeqTraceCons(new Integer( state ),
                                      cf.ignoreValue( elementType ),
                                      _ref( accumSym ));
      }

}

File: sources/scala/tools/scalac/transformer/matching/LeftTracerInScala.scala
import scalac._;
import scalac.ast._;
import scalac.symtab._;
import Tree._;
import scalac.util.Names;

import java.util._ ;

import scala.tools.util.Position;
//import scalac.transformer.matching._ ;
import scalac.transformer.matching.CodeFactory ;
import scalac.transformer.matching.DetWordAutom ;
import scalac.transformer.matching.Label ;
import scalac.transformer.matching.PartialMatcher ;
package scala.tools.scalac.transformer.matching {

class LeftTracerInScala(dfa: DetWordAutom, elementType: Type, owner: Symbol, cf: CodeFactory, val selector: Tree )
extends TracerInScala( dfa, elementType, owner, cf ) {

  final def defs = cf.defs;

    /** symbol of the accumulator ( scala.SequenceList )
     */
    var accumSym: Symbol   = _;
    var accumType: Type    = _;
    var accumTypeArg: Type =_ ;

    def _accumType(elemType: Type): Type =  {
        cf.SeqTraceType( elemType );
    }


    protected def initializeSyms(): Unit = {
      this.funSym =  owner.newLabel( pos,
                                    cf.fresh.newName( "left" ));

      this.iterSym = owner.newVariable( pos,
                                       Modifiers.MUTABLE,
                                       cf.fresh.newName( "iter" ))
      .setType( cf._seqIterType( elementType ) ) ;

      this.stateSym = owner.newVariable( pos,
                                        Modifiers.MUTABLE,
                                        cf.fresh.newName( "q" ))
      .setType( defs.int_TYPE() ) ;

      this.accumType = _accumType( elementType );
      this.accumTypeArg = accumType.typeArgs()( 0 );
      this.accumSym = owner.newVariable( pos,                  // accumulator
                                        Modifiers.MUTABLE,
                                        cf.fresh.newName( "acc" ))
      .setType( accumType );

      //this.funSym
      //    .setType( new Type.MethodType( new Symbol[] {
      //        accumSym, iterSym, stateSym},
      //                                   accumType));

      this.funSym
      .setType( new Type.MethodType( Predef.Array[Symbol] (  // dummy symbol MethodType
        funSym.newVParam( pos, 0, cf.fresh.newName( "q" ), defs.int_TYPE()),
        funSym.newVParam( pos, 0, cf.fresh.newName( "acc" ), accumType ) ),
                                    accumType)); // result type = List[T]

      this.resultSym = owner.newVariable(pos,
                                         0,
                                         cf.fresh.newName("trace"))
      .setType( accumType ) ;

      this.curSym = owner.newVariable( pos, 0, Names.cur )
      .setType( elementType );

      this.hasnSym = owner.newVariable( pos, 0, Names.hasNext )
      .setType( defs.boolean_TYPE() );

    }

  /* should throw an exception here really, e.g. MatchError
   */
  override def code_fail() = {
     gen.Ident( accumSym.pos, accumSym );
  }

  /** returns translation of transition with label from i.
     *  returns null if there is no such transition(no translation needed)
     */
    override def code_delta(i: Int, label: Label): Tree = {
      val target = dfa.delta( i, label );

      /*
       System.out.println("LeftTracer:calling dfa.delta("+i+","+label+")");
       System.out.println("result: "+target);
       */
      if( target == null )
        null;
      else {
        // (optimization) that one is a dead state (does not make sense for tracer)
        /*
          if( target == dfa.nstates - 1 )
          return code_fail();
        */

        /*
          Tree newAcc = cf.newSeqTraceCons(new Integer(i),
          currentElem(),
          _ref( accumSym ));
        */
        val hd = cf.newPair( gen.mkIntLit(cf.pos, i), currentElem() );
        val newAcc = gen.mkNewCons( cf.pos,
                                   accumTypeArg,
                                   hd,
                                   gen.Ident( cf.pos, accumSym ));

        //return callFun( new Tree[] { newAcc , _iter(), gen.mkIntLit( cf.pos, target )} );
        callFun( Predef.Array[Tree]( gen.mkIntLit( cf.pos, target.intValue() ), newAcc ) );
      }
    }


  def code_body(): Tree = {

    var body = code_error(); // never reached at runtime.

    // state [ nstates-1 ] is the dead state, so we skip it

    //`if( state == q ) <code_state> else {...}'
    var i = dfa.nstates()-2;
    while(i >= 0) {
      body = code_state( i, body );
      i = i - 1;
    }
    loadCurrentElem( body );
  }

    /** return code for state i of the dfa SAME AS IN SUPER, ONLY SINK IS GONE
     */
  def code_state(i: Int, elseBody: Tree): Tree  = {

    var runFinished: Tree = _; // holds result of the run
    var finalSwRes: Int = 0;

    runFinished = run_finished( i );

    var stateBody: Tree = _ ; // action(delta) for one particular label/test

    // default action (fail if there is none)

    stateBody = code_delta( i, Label.DefaultLabel);

    if( stateBody == null )
      stateBody = code_fail();
    // transitions of state i

    val trans = (dfa.deltaq.asInstanceOf[Array[HashMap]])( i );
    val labs = (dfa.deltaq( i ).asInstanceOf[HashMap]).keySet().iterator();
    while(labs.hasNext()) {
      val label = labs.next();
      val next = trans.get( label ).asInstanceOf[Integer];

      val  action = code_delta( i, label.asInstanceOf[Label] );

      if( action != null ) {
        stateBody = gen.If( currentMatches(label.asInstanceOf[Label]),
                           action,
                           stateBody);
      }
    }
    stateBody = gen.If( cf.Negate( gen.Ident( cf.pos, hasnSym )),
                       runFinished,
                       stateBody );
    gen.If( cf.Equals( _state(), gen.mkIntLit(cf.pos, i )),
           stateBody ,
           elseBody );
  }

  def  getTrace(): Tree = {

    initializeSyms();

    cf.gen.mkBlock( cf.pos, Predef.Array[Tree] (
      gen.ValDef( iterSym, cf.newIterator( selector, selector.getType() )),
      gen.ValDef( stateSym, gen.mkIntLit( cf.pos, 0) ),
      gen.ValDef( accumSym, gen.mkNil( cf.pos )),
      gen.ValDef( resultSym,
                 gen.LabelDef( this.funSym,
                              Predef.Array[Ident] (
                                gen.Ident( pos, stateSym ),
                                gen.Ident( pos, accumSym )
                              ), code_body() /* code_body_new ? */ ))
    ),
                   gen.Ident( cf.pos, resultSym ));
  }

  // calling the AlgebraicMatcher here
  override def _cur_match(pat: Tree): Tree  = {
    //return gen.mkBooleanLit(cf.pos, true);

    //System.out.println("calling algebraic matcher on type:"+pat.type);

    val m = new PartialMatcher( owner,
                               currentElem(),
                               defs.boolean_TYPE() );

    val res1 = if(containsBinding( pat )) {
      pat.match {
        case Sequence(pats) =>
          gen.mkBooleanLit(cf.pos, true);
        case _ =>
          null
      }
    } else null;

    if (res1 == null) {

      am.construct( m, Predef.Array[Tree] (
        cf.gen.CaseDef( pat,
                       gen.mkBooleanLit( cf.pos, true )),
        cf.gen.CaseDef( cf.gen.Ident( pat.pos, defs.PATTERN_WILDCARD ),
                       gen.mkBooleanLit( cf.pos, false)) ),
                   false);
      val res = am.toTree();
      // debugprint ?
      res;
    } else null
  }


    /** return the accumulator + last state
     */
    override def run_finished(state: Int): Tree  = {
      val hd = cf.newPair( gen.mkIntLit( cf.pos, state ),
                          gen.mkDefaultValue(cf.pos,
                                             elementType));
      //System.err.println(hd.type);
      gen.mkNewCons(  cf.pos,
                    accumTypeArg,
                    hd,
                    gen.Ident( cf.pos, accumSym ));
    }

}
}
